api/valor_api/backend/core/datum.py

```diff
@@ -36,12 +44,58 @@ def create_datums(
     ]
 
     try:
-        db.add_all(rows)
+        if ignore_existing_datums:
+            values = [
+                {
+                    "uid": row.uid,
+                    "dataset_id": row.dataset_id,
+                    "meta": row.meta,
+                }
+                for row in rows
+            ]
+            insert_stmt = (
+                insert(models.Datum)
+                .values(values)
+                .on_conflict_do_nothing(index_elements=["dataset_id", "uid"])
+            )
+
+            db.execute(insert_stmt)
+
+            # get all the ids and update the row objects
+            for row in rows:
+                row.id = (
```

Comment:
Instead of iterating through all the rows again to retrieve ids you could try calling `.returning` on the insert statement.

See [reference](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#getting-new-objects-with-returning).

```
insert_stmt = (
    insert(models.Datum)
    .values(values)
    .on_conflict_do_nothing(index_elements=["dataset_id", "uid"])
    .returning(models.Datum)
)
rows = db.scalars(insert_stmt).all()
```

Reply:
i tried that but it will only return the new ones and we need all of them

Reply:
Im not sure whether this will miss the conflicted datums

Reply:
oh wait i guess actually we only want to add annotations to the newly created ones.....

Reply:
I think youre right b/c the create annotations uses a zip which requires the ordered list...

```
for annotations_per_datum, datum, model in zip(
            annotations, datums, models_
        )
```

This should never be run if there exist duplicates